### ISSUE TITLE
Allow functions calls in `goverter:map`

### ISSUE THREAD

**Issue by s1berc0de** [jmattheis/goverter#91]
I think it would be a great idea to map source structure method result to a target structure field (it can help to map private structure fields)

**Comment by jmattheis** [comment]
Could you give an example?

**Comment by s1berc0de** [comment]
Smth like this

type Input struct {
 id uint64
}

func (i Input) ID() uint64 {
 return i.id
}

type Output struct {
 ID uint64
}

// goverter:converter
type Converter interface {
  // goverter:map ID() | ID
  Convert(source Input) Output
}

**Comment by jmattheis** [comment]
Yeah, sounds good. I'd say goverter should automatically try to map methods to properties, so your example above should be automatically handled. Still it should be possible to manually define it, when the names on the source and target doesn't match
```go
type Input struct {
 id uint64
}

func (i Input) Identifier() uint64 {
 return i.id
}

type Output struct {
 ID uint64
}

// goverter:converter
type Converter interface {
  // goverter:map Identifier() ID
  Convert(source Input) Output
}
```

**Comment by s1berc0de** [comment]
Exactly, method and field can't have the same names, so if goverter can't find struct field with this name, it should to search a method, and now we have one restriction that is one method parameter, I think an idea to convert multi inputs in one output also good, it's pretty good if you use domain aggreagate of few entities

**Comment by jmattheis** [comment]
> now we have one restriction that is one method parameter, I think an idea to convert multi inputs in one output also good, it's pretty good if you use domain aggreagate of few entities

Could you elaborate this with an example?

Goverter should be able to map fields on anonymous structs, but the definition is a bit cumbersome, could you give an example that doesn't work?

**Comment by s1berc0de** [comment]
About multi inputs:
type EntityMixin struct {
	CreatedAt time.Time
	UpdatedAt time.Time
	DeletedAt *time.Time
	ID        uint64
}

type Account struct {
	Version     time.Time
	ConfirmedAt sql.NullTime
	Email       string
	Password    string
	EntityMixin
	Status enums.AccountStatus
}

type Role struct {
	ID   uint64
	Type enums.RoleType
}

type Group struct {
	gorm.EntityMixin
	Name string
}

// goverter:converter
type Converter interface {
	// goverter:map accountEntity.EntityMixin.ID ID
	// goverter:map accountEntity.Email              Email
	// goverter:map accountEntity.Version           Version
	// goverter:map roleEntity.Type                      RoleType
	// goverter:map groupEntity.EntityMixin.ID    GroupID
	DatabaseToEntities(accountEntity *entities.Account, roleEntity *entities.Role, groupEntity *entities.Group) *identity.Identity
}

**Comment by s1berc0de** [comment]
About one input, but nested EntityMixin struct, entities the same as upper, but contract is:

type DatabaseToEntitiesInput struct {
	AccountEntity *entities.Account
	RoleEntity    *entities.Role
	GroupEntity   *entities.Group
}

// goverter:converter
type Converter interface {
	// goverter:map AccountEntity.EntityMixin.ID ID
	// goverter:map AccountEntity.Email              Email
	// goverter:map AccountEntity.Version           Version
	// goverter:map RoleEntity.Type                     RoleType
	// goverter:map GroupEntity.EntityMixin.ID    GroupID
	DatabaseToEntities() *identity.Identity
}

Error while creating converter method: ....

|     |
|     |
|     |
source*.???
target*.ID
|     | |
|     | | uint64
|     |
...

Cannot match the target field with the source entry: "ID" does not exist.
exit status 1

**Comment by jmattheis** [comment]
Multiple inputs will probably be supported with #68.

Do you mean in the second example `DatabaseToEntities(*DatabaseToEntitiesInput) *identity.Identity`?

If yes, then you can use autoMap for this
```
// goverter:converter
type Converter interface {
	// goverter:autoMap AccountEntity
	// goverter:autoMap RoleEntity
	// goverter:autoMap GroupEntity
	DatabaseToEntities(*DatabaseToEntitiesInput) *identity.Identity
}
```
See https://goverter.jmattheis.de/#/conversion/mapping?id=auto-map & https://github.com/jmattheis/goverter/issues/55

**Comment by s1berc0de** [comment]
autoMap doesn't work because of anonymous nested EntityMixin

**Comment by jmattheis** [comment]
Can you provide a smaller example that is self contained? You should be able to access the anonymous or rather embedded struct by the type name in your case EntityMixin in goverter. See https://github.com/jmattheis/goverter/issues/24

**Comment by s1berc0de** [comment]
Yes, if i write the mapping functuin i can get access to the ID field through the EntityMixin, but goverter doesn't see this ability

**Comment by s1berc0de** [comment]
type A struct {
	ID uint64
}

type B struct {
	A
	Test string
}

type C struct {
	ID   uint64
	Test string
}

type DatabaseToEntitiesInput struct {
	B B
}

// goverter:converter
type Converter interface {
	// goverter:autoMap B
	DatabaseToEntities(*DatabaseToEntitiesInput) *C
}

Gets error too

**Comment by s1berc0de** [comment]
Oh, if i don't use pointers inside of DatabaseToEntitiesInput it can map field from embedded struct

**Comment by jmattheis** [comment]
ID is on the A struct so it isn't seen by just automapping B, you need properties from both B and B.A: 
```go
// goverter:converter
type Converter interface {
	// goverter:autoMap B.A
	// goverter:autoMap B
	DatabaseToEntities(*DatabaseToEntitiesInput) *C
}
```
This builds without problems.

**Comment by s1berc0de** [comment]
Yeah, got now problems with mapping time.Time type 

|     | |
|     | |             | time.Time
|     | |             |
|     | |             |
|     | |             |
source*.AccountEntity.Version.???
target*              .Version.wall
|     |               |       |
|     |               |       | uint64
|     |               |
|     |               | time.Time
|     |
|     |
|     |

**Comment by s1berc0de** [comment]
type A struct {
	ID uint64
}

type B struct {
	A
	Version time.Time
	Test    string
}

type C struct {
	ID      uint64
	Version time.Time
	Test    string
}

type DatabaseToEntitiesInput struct {
	B B
}

// goverter:converter
type Converter interface {
	// goverter:map B.A.ID      ID
	// goverter:map B.Test               Test
	// goverter:map B.Version Version
	DatabaseToEntities(*DatabaseToEntitiesInput) *C
}

**Comment by jmattheis** [comment]
Goverter doesn't know that see https://github.com/jmattheis/goverter/issues/4

**Comment by s1berc0de** [comment]
Yeah i se, thnq for the discussion, so now we have 2 features: #68 and that is in the title

**Comment by jmattheis** [comment]
Feature added in v1.1.0. Example:

Input:
```go
package structs

// goverter:converter
type Converter interface {
    Convert(source Input) Output
}

type Input struct {
    Name string
}
func (Input) Age() int {
    return 42
}
type Output struct {
    Name string
    Age int
}
```
Output:
```go
// Code generated by github.com/jmattheis/goverter, DO NOT EDIT.

package generated

import execution "github.com/jmattheis/goverter/execution"

type ConverterImpl struct{}

func (c *ConverterImpl) Convert(source execution.Input) execution.Output {
    var structsOutput execution.Output
    structsOutput.Name = source.Name
    structsOutput.Age = source.Age()
    return structsOutput
}
```
Fixes #91
